lib/check-meta.nix

```diff
@@ -0,0 +1,200 @@
+let
+  lib = import ./default.nix;
+in rec {
+
+  # Extend a derivation with checks for brokenness, license, etc.
+  # Throw a descriptive error when the check fails; return derivaitonArg otherwise.
+  # Note: no dependencies are checked in this step.
+  addMetaChecks = { config, platform, meta, derivationArg, mkDerivationArg }: let
+    attrs = mkDerivationArg; # TODO: probably get rid of passing this one
+
+    # See discussion at https://github.com/NixOS/nixpkgs/pull/25304#issuecomment-298385426
+    # for why this defaults to false, but I (@copumpkin) want to default it to true soon.
+    shouldCheckMeta = config.checkMeta or false;
+
+    allowUnfree = config.allowUnfree or false || builtins.getEnv "NIXPKGS_ALLOW_UNFREE" == "1";
+
+    whitelist = config.whitelistedLicenses or [];
+    blacklist = config.blacklistedLicenses or [];
+
+    onlyLicenses = list:
+      lib.lists.all (license:
+        let l = lib.licenses.${license.shortName or "BROKEN"} or false; in
+        if license == l then true else
+          throw ''‘${showLicense license}’ is not an attribute of lib.licenses''
+      ) list;
+
+    areLicenseListsValid =
+      if lib.mutuallyExclusive whitelist blacklist then
+        assert onlyLicenses whitelist; assert onlyLicenses blacklist; true
+      else
+        throw "whitelistedLicenses and blacklistedLicenses are not mutually exclusive.";
+
+    hasLicense = attrs:
+      attrs ? meta.license;
+
+    hasWhitelistedLicense = assert areLicenseListsValid; attrs:
+      hasLicense attrs && builtins.elem attrs.meta.license whitelist;
+
+    hasBlacklistedLicense = assert areLicenseListsValid; attrs:
+      hasLicense attrs && builtins.elem attrs.meta.license blacklist;
+
+    allowBroken = config.allowBroken or false || builtins.getEnv "NIXPKGS_ALLOW_BROKEN" == "1";
+
+    isUnfree = licenses: lib.lists.any (l:
+      !l.free or true || l == "unfree" || l == "unfree-redistributable") licenses;
+
+    # Alow granular checks to allow only some unfree packages
+    # Example:
+    # {pkgs, ...}:
+    # {
+    #   allowUnfree = false;
+    #   allowUnfreePredicate = (x: pkgs.lib.hasPrefix "flashplayer-" x.name);
+    # }
+    allowUnfreePredicate = config.allowUnfreePredicate or (x: false);
```

Comment:
Would it make sense to also allow a blacklist implementation? Right now it's only possible to go from false -> true.